foreshadow/tests/test_core/test_fs_logging.py

```diff
@@ -0,0 +1,221 @@
+
+
+@pytest.mark.parametrize(
+    "level", ["debug", "info", "warning", "error", "critical"]
+)
+def test_get_log_fn(caplog, level):
+    """Test that log_fn logs properly to a stream.
+
+    Args:
+        caplog: captures logging output.
+        level: the loggig level to test
+
+    """
+    from foreshadow.core import fs_logging
+
+    logger = fs_logging.get_logger()
+    logger.setLevel(fs_logging.LEVELS[level])
+    test_print = "testing"
+    log_fn = fs_logging._get_log_fn(level)
+    log_fn(test_print)
+    assert caplog.record_tuples[0][2] == test_print
+
+
+@pytest.mark.parametrize(
+    "level,schema,outfile",
+    [
+        ("debug", "MetricSchema", "gui_data.txt"),
+        ("info", "MetricSchema", "gui_data.txt"),
+        ("warning", "MetricSchema", "gui_data.txt"),
+        ("error", "MetricSchema", "gui_data.txt"),
+        ("critical", "MetricSchema", "gui_data.txt"),
+        ("debug", "MetricSchema", "random_file.txt"),
+        ("debug", "MetricSchema", "gui_data.txt"),
+    ],
+)
+@mock.patch("builtins.open", new_callable=mock.mock_open())
+def test_log_and_gui(mock_open, caplog, level, schema, outfile):
+    """Test log_and_gui logs and writes to gui.
+
+    Args:
+        mock_open: mock the open builtin call and test it being called
+        caplog: captures logging output.
+        level: the loggig level to test
+        schema: schema to use for gui
+        outfile: outfile to write to.
+
+    """
+    from foreshadow.core import fs_logging
+    from foreshadow.tests.utils import dynamic_import
+
+    fs_logging.gui_fn.buffer_size = 1
+    fs_logging.gui_fn.first_write = True  # have to set as each parametrize
+    # will use the previous call's gui_fn, meaning first_write will already
+    # be set to False
+    fs_logging.gui_fn.outfile = outfile
+    path = "foreshadow.core.gui"
+    schema = dynamic_import(schema, path)()
+    schema_fields = schema.declared_fields
+    details = {}
+    for name, field in schema_fields.items():
+        field_type = type(field).__name__.split(".")[-1]
+        if field_type == "String":
+            func = str
+        elif field_type == "Float":
+            func = float
+        else:
+            raise NotImplementedError(
+                "field type: %s is not valid" % field_type
+            )
+        details[name] = func(1)
+    fs_logging.set_level("debug")
+    msg = "test"
+    fs_logging.log_and_gui(level, msg, schema, details)
+    # msg = level.upper()+':'+'foreshadow:'+msg
+    assert caplog.record_tuples[0][2] == msg
+    mock_open.assert_called_with(fs_logging.gui_fn.outfile, "w")
+
+
+@pytest.mark.parametrize(
+    "level", ["debug", "info", "warning", "error", "critical"]
+)
+def test_set_level(level):
+    """ Test ability to set logging level.
+
+    Args:
+        level: logging level to set.
+
+    """
+    from foreshadow.core import fs_logging
+
+    fs_logging.set_level(level)
+    assert fs_logging.get_logger().level == fs_logging.LEVELS[level]
```

Comment:
Having two convenience global vars might be useful much like [pickle](https://docs.python.org/3/library/pickle.html#pickle.HIGHEST_PROTOCOL)

Reply:
We could do that if you want, I don't see how it would be easier than calling set_level('debug') or set_level('info') though?